### PR TITLE
[Api][Shop] Enable order resource

### PR DESCRIPTION
**Pull request by Wojdylak**
| Q               | A
|-----------------|-----
| Branch?         | api-platform-3
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | n/a
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Resources/config/suites/api.yml [hunk 1]
```diff
@@ -56,13 +56,13 @@ imports:
     - api/promotion/managing_promotions.yml
 #    - api/promotion/receiving_discount.yml
     - api/promotion/removing_catalog_promotions.yml
-#    - api/shipping/applying_shipping_method_rules.yml TODO: Enable after making checkout resource available in API
-#    - api/shipping/applying_shipping_fee.yaml TODO: Enable after making checkout resource available in API
-#    - api/shipping/viewing_shipping_methods.yml TODO: Enable after making checkout resource available in API
+    - api/shipping/applying_shipping_method_rules.yml
+#    - api/shipping/applying_shipping_fee.yml TODO: Enable after making checkout resource available in API
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Resources/config/suites/api.yml:60]: What about this one?

File: tests/Api/Responses/shop/order/get_configuration.json [hunk 2]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (NoResponseMate) [inline tests/Api/Responses/shop/order/get_configuration.json]: Would be nice to have something here

File: src/Sylius/Behat/Resources/config/suites/api.yml (hunk 1 above)
Comment (Wojdylak) [reply src/Sylius/Behat/Resources/config/suites/api.yml:60]: This suit will be enabled in the next PR

File: features/shop/checkout/paying_for_order/paying_offline_during_checkout_as_guest.feature [hunk 3]
```diff
@@ -20,15 +20,15 @@ Feature: Paying Offline during checkout as guest
...
+    @no-api @ui
     Scenario: Successfully placing an order using custom locale
         Given I have product "PHP T-Shirt" in the cart
         When I proceed through checkout process in the "French (France)" locale with email "john@example.com"
         And I confirm my order
         Then I should see the thank you page in "French (France)"
 
     @api @no-ui
-    Scenario: Successfully placing an order using custom locale
+    Scenario: Successfully placing an order using custom locale2
```
Comment (Rafikooo) [inline features/shop/checkout/paying_for_order/paying_offline_during_checkout_as_guest.feature]:
```suggestion
    Scenario: Successfully placing an order using the selected custom locale during cart pickup
```
wdyt?

File: features/shop/checkout/paying_for_order/paying_with_paypal_during_checkout.feature [hunk 4]
```diff
@@ -23,7 +23,7 @@ Feature: Paying with paypal during checkout
         And I should see the thank you page
         And the latest order should have a payment with state "completed"
 
-    @ui
+    @no-api @ui
```
Comment (Rafikooo) [inline features/shop/checkout/paying_for_order/paying_with_paypal_during_checkout.feature:26]: Should these be covered in API as well?

File: features/shop/checkout/shipping_order/preventing_not_available_shipping_method_selection.feature [hunk 5]
```diff
@@ -87,7 +87,7 @@ Feature: Preventing not available shipping method selection
         And I complete the addressing step
         Then I should not be able to select "Dragon Post" shipping method
 
-    @api
+    @api @no-ui
     Scenario: Not being able to select inexistent shipping method
```
Comment (Rafikooo) [inline features/shop/checkout/shipping_order/preventing_not_available_shipping_method_selection.feature]:
```suggestion
    @api @no-ui
    Scenario: Not being able to select nonexistent shipping method
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 6]
```diff
@@ -1168,7 +1179,7 @@ public function iShouldNotBeAbleToResendTheOrderConfirmationEmail(): void
         $this->client->customItemAction(
             Resources::ORDERS,
             $this->sharedStorage->get('order')->getTokenValue(),
-            HttpRequest::METHOD_POST,
+            HttpRequest::METHOD_PATCH,
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: Why has this operation method been changed?
https://github.com/Sylius/Sylius/pull/16666/files#diff-26c665fdaf982c8a1dc0c824ab126511cd2b7031ab88bc67304a323aee732871R59

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 7]
```diff
@@ -398,7 +396,7 @@ public function iTryToSelectShippingMethod(string $shippingMethodCode): void
             HTTPRequest::METHOD_PATCH,
             sprintf('shipments/%s', $this->getCart()['shipments'][0]['id']),
         );
-        $request->setContent(['shippingMethod' => $this->iriConverter->getItemIriFromResourceClass($this->shippingMethodClass, ['code' => $shippingMethodCode])]);
+        $request->setContent(['shippingMethod' => sprintf('/api/v2/shop/shipping-methods/%s', $shippingMethodCode)]);
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:399]: Just a general thought - since you removed the usage of IRI converter here, what about removing its usage throughout the entire file and eliminating the service?

File: src/Sylius/Behat/Context/Setup/CheckoutContext.php [hunk 8]
```diff
@@ -73,8 +72,7 @@ public function iHaveProceededThroughCheckoutProcess(): void
         $cart = $this->orderRepository->findCartByTokenValue($cartToken);
         Assert::notNull($cart);
 
-        $command = new UpdateCart(null, $this->getDefaultAddress());
-        $command->setOrderTokenValue($cartToken);
+        $command = new UpdateCart(null, $this->getDefaultAddress(), orderTokenValue: $cartToken);
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Setup/CheckoutContext.php]: It would be more informative with only the used named arguments.

File: src/Sylius/Bundle/ApiBundle/Command/Checkout/CompleteOrder.php [hunk 9]
```diff
@@ -17,24 +17,14 @@
...
-    public function __construct(?string $notes = null)
-    {
-        $this->notes = $notes;
+    public function __construct(
+        public ?string $notes = null,
+        public ?string $orderTokenValue = null,
+    ) {
     }
 
     public function getOrderTokenValue(): ?string
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Command/Checkout/CompleteOrder.php:26]: Considering that commands can be read-only with the introduction of the context builder, what are your thoughts on removing these command getters and the interface method?

File: src/Sylius/Bundle/ApiBundle/Command/Checkout/ChoosePaymentMethod.php [hunk 10]
```diff
@@ -15,64 +15,30 @@
...
-    public $paymentMethodCode;
-
-    public function __construct(string $paymentMethodCode)
-    {
-        $this->paymentMethodCode = $paymentMethodCode;
+    public function __construct(
+        public string $paymentMethodCode,
+        public ?int $paymentId = null,
+        public ?string $orderTokenValue = null,
+    ) {
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Command/Checkout/ChoosePaymentMethod.php:27]:
```suggestion
    public function __construct(
        public string $orderTokenValue,
        public int $paymentId,
        public string $paymentMethodCode,
    ) {
```
1. IMO it's a more natural order.
2. Should the command now be strictly typed since we create them with all the necessary data?

File: src/Sylius/Bundle/ApiBundle/Command/PaymentIdAwareInterface.php [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Command;
+
+interface PaymentIdAwareInterface extends CommandAwareDataTransformerInterface
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Command/PaymentIdAwareInterface.php:16]: I have some concerns regarding the validity of the `CommandAwareDataTransformerInterface`:

The interface can be found here:
[CommandAwareDataTransformerInterface](https://github.com/Sylius/Sylius/blob/f6a3f436204c2b143733f88a2657bb48f612a4b3/src/Sylius/Bundle/ApiBundle/Command/CommandAwareDataTransformerInterface.php#L16)

It is only used within this class:
[CommandAwareInputDataTransformer](https://github.com/Sylius/Sylius/blob/f6a3f436204c2b143733f88a2657bb48f612a4b3/src/Sylius/Bundle/ApiBundle/DataTransformer/CommandAwareInputDataTransformer.php#L44)

From my observation, this class implements the `ApiPlatform\Core\DataTransformer\DataTransformerInterface`, which was removed in API Platform 3. Given this, I'm wondering if both the CommandAwareInputDataTransformer and the broader use of the DataTransformer concept are still valid within the context of API Platform 3.

Additionally, I noticed a potential issue with this compiler pass:
[CommandDataTransformerPass](https://github.com/Sylius/Sylius/blob/f6a3f436204c2b143733f88a2657bb48f612a4b3/src/Sylius/Bundle/ApiBundle/DependencyInjection/Compiler/CommandDataTransformerPass.php#L26)

It seems this compiler pass might be obsolete because it adds the following deprecated tag:
`$commandDataTransformersChainDefinition->addTag('api_platform.data_transformer');`

When debugging, the line:
`$taggedServices = $container->findTaggedServiceIds('sylius.api.command_data_transformer');`
returns no services, but in API Platform 2.7 the following ones:
```
array:8 [
  0 => "Sylius\Bundle\ApiBundle\DataTransformer\OrderTokenValueAwareInputCommandDataTransformer"
  1 => "Sylius\Bundle\ApiBundle\DataTransformer\ShipmentIdAwareInputCommandDataTransformer"
  2 => "Sylius\Bundle\ApiBundle\DataTransformer\LoggedInShopUserIdAwareCommandDataTransformer"
  3 => "Sylius\Bundle\ApiBundle\DataTransformer\ChannelCodeAwareInputCommandDataTransformer"
  4 => "Sylius\Bundle\ApiBundle\DataTransformer\LoggedInCustomerEmailIfNotSetAwareCommandDataTransformer"
  5 => "Sylius\Bundle\ApiBundle\DataTransformer\LoggedInCustomerEmailAwareCommandDataTransformer"
  6 => "Sylius\Bundle\ApiBundle\DataTransformer\LocaleCodeAwareInputCommandDataTransformer"
  7 => "Sylius\Bundle\ApiBundle\DataTransformer\SubresourceIdAwareCommandDataTransformer"
]
```

I found only one occurrence of this ID:
[AsCommandDataTransformer](https://github.com/Sylius/Sylius/blob/f6a3f436204c2b143733f88a2657bb48f612a4b3/src/Sylius/Bundle/ApiBundle/Attribute/AsCommandDataTransformer.php#L19)

However, it seems that these attributes are registered differently:
[SyliusApiExtension](https://github.com/Sylius/Sylius/blob/f6a3f436204c2b143733f88a2657bb48f612a4b3/src/Sylius/Bundle/ApiBundle/DependencyInjection/SyliusApiExtension.php#L74)

To sum up, while I'm uncertain if my concerns are entirely accurate, I wanted to raise these points for your insight and analysis.

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml [hunk 12]
```diff
@@ -17,19 +17,15 @@
...
             <group>sylius:shop:cart:update</group>
         </attribute>
         <attribute name="shippingAddress">
-            <group>shop:cart:update</group>
             <group>sylius:shop:cart:update</group>
         </attribute>
         <attribute name="email">
-            <group>shop:cart:update</group>
-            <group>sylius:shop:cart:update</group>
+            <group>sylius:shop:guest:write</group>
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml:26]:
```suggestion
            <group>sylius:shop:guest:update</group>
```
Wouldn't it be more predictable to stick to commonly used groups, wdyt?

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml [hunk 13]
```diff
@@ -17,19 +17,15 @@
 >
     <class name="Sylius\Bundle\ApiBundle\Command\Checkout\UpdateCart">
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml:18]: Another general thought:
While we have action-specific commands like `ChooseShippingMethod` or `ChoosePaymentMethod`, this command is less descriptive. Upon reviewing the handler, it actually performs three tasks: updating addresses, applying coupons, and setting the order owner's email. Wouldn't it be better to split it into three smaller, action-specific commands? I understand this would mean adding more endpoints, but what do you think?

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/command_handlers.xml [hunk 14]
```diff
@@ -58,10 +58,10 @@
             <tag name="messenger.message_handler" bus="sylius_default.bus" />
         </service>
 
-<!--        <service id="Sylius\Bundle\ApiBundle\CommandHandler\Cart\InformAboutCartRecalculationHandler">-->
-<!--            <tag name="messenger.message_handler" bus="sylius.command_bus" />-->
-<!--            <tag name="messenger.message_handler" bus="sylius_default.bus" />-->
-<!--        </service>-->
+        <service id="Sylius\Bundle\ApiBundle\CommandHandler\Cart\InformAboutCartRecalculationHandler">
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/command_handlers.xml:61]: Since handlers are services, shouldn't they use dotted ID formatting as well?

File: src/Sylius/Bundle/ApiBundle/Command/Account/ChangePaymentMethod.php [hunk 15]
```diff
@@ -15,45 +15,24 @@
...
-     * @var string|null
-     */
-    public $paymentId;
-
-    public function __construct(public string $paymentMethodCode)
-    {
+    public function __construct(
+        public string $paymentMethodCode,
+        public ?int $paymentId = null,
+        public ?string $orderTokenValue = null,
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Command/Account/ChangePaymentMethod.php:25]:
```suggestion
        public string $orderTokenValue,
        public int $paymentId,
        public string $paymentMethodCode,
```
🤔

File: src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Shop/Order/StateBasedExtension.php [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+use Sylius\Bundle\ApiBundle\SectionResolver\ShopApiSection;
+use Sylius\Bundle\CoreBundle\SectionResolver\SectionProviderInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+
+final readonly class StateBasedExtension implements QueryCollectionExtensionInterface, QueryItemExtensionInterface
+{
+    /** @param array<string> $nonFilteredCartAllowedOperations */
+    public function __construct(
+        private SectionProviderInterface $sectionProvider,
+        private array $nonFilteredCartAllowedOperations = [],
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Shop/Order/StateBasedExtension.php]: I suggest either:
```suggestion
        private array $cartAllowedOperations = [],
        or
        private array $allowedOperations = [],
```
I don't see the value in the `nonFiltered` prefix.

File: features/shop/checkout/ability_to_change_email_during_checkout.feature [hunk 17]
```diff
@@ -11,15 +11,15 @@ Feature: Changing email during checkout with registered email
         And the store allows paying Offline
         And there is a customer "John Doe" identified by an email "john@example.com" and a password "secret"
 
-    @ui
+    @api-todo @ui #TODO: Checkout
```
Comment (TheMilek) [inline features/shop/checkout/ability_to_change_email_during_checkout.feature:14]: why? 🤔

File: features/shop/checkout/paying_for_order/paying_offline_during_checkout_as_guest.feature (hunk 3 above)
Comment (TheMilek) [inline features/shop/checkout/paying_for_order/paying_offline_during_checkout_as_guest.feature]:
```suggestion
    Scenario: Successfully placing an order using custom locale
```

File: features/shop/checkout/paying_for_order/paying_with_paypal_during_checkout.feature (hunk 4 above)
Comment (TheMilek) [inline features/shop/checkout/paying_for_order/paying_with_paypal_during_checkout.feature:26]: just wondering if we shouldn't wait for the payment request to be sure to add this tag 🤔

File: features/shop/checkout/shipping_order/preventing_not_available_shipping_method_selection.feature (hunk 5 above)
Comment (TheMilek) [inline features/shop/checkout/shipping_order/preventing_not_available_shipping_method_selection.feature]:
```suggestion
    Scenario: Not being able to select nonexistent shipping method
```

File: features/shop/checkout/shipping_order/returning_from_shipping_step_to_addressing_step.feature [hunk 18]
```diff
@@ -10,15 +10,15 @@ Feature: Returning from shipping step to addressing step
         And the store ships everywhere for Free
         And I am a logged in customer
 
-    @ui @no-api
+    @no-api @ui
     Scenario: Going back to addressing step with button
         Given I have product "Apollo 11 T-Shirt" in the cart
         And I specified the billing address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
         And I decide to change my address
```
Comment (TheMilek) [inline features/shop/checkout/shipping_order/returning_from_shipping_step_to_addressing_step.feature]:
```suggestion
        When I decide to change my address
```

Comment (TheMilek) [reply features/shop/checkout/shipping_order/returning_from_shipping_step_to_addressing_step.feature]: same below

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php (hunk 7 above)
Comment (TheMilek) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:399]:
```suggestion
       $request->setContent(['shippingMethod' => $this->iriConverter->getItemIriFromResourceClass($shippingMethod)]);
```
I would use transform to retrieve the shippingMethod directly and use it here

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 19]
```diff
@@ -408,14 +406,15 @@ public function iTryToSelectShippingMethod(string $shippingMethodCode): void
...
         $request = $this->requestFactory->customItemAction(
             'shop',
             Resources::ORDERS,
             $this->sharedStorage->get('cart_token'),
             HTTPRequest::METHOD_PATCH,
-            sprintf('payments/%s', $this->getCart()['payments'][0]['id']),
+            sprintf('payments/%s', $cart['payments'][0]['id']),
         );
-        $request->setContent(['paymentMethod' => $this->iriConverter->getItemIriFromResourceClass($this->paymentMethodClass, ['code' => $paymentMethodCode])]);
+        $request->setContent(['paymentMethod' => sprintf('/api/v2/shop/payment-methods/%s', $paymentMethodCode)]);
```
Comment (TheMilek) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:417]: same here as above

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 20]
```diff
@@ -408,14 +406,15 @@ public function iTryToSelectShippingMethod(string $shippingMethodCode): void
      */
     public function iTryToSelectPaymentMethod(string $paymentMethodCode): void
```
Comment (TheMilek) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:407]:
```suggestion
    public function iTryToSelectPaymentMethod(string $paymentMethod): void
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 21]
```diff
@@ -1163,7 +1162,7 @@ public function iTryToAddProductVariantWithCode(ProductInterface $product, strin
         /** @var ProductVariantInterface $variant */
         $variant = $product->getVariants()->first();
         $request->setContent([
-            'productVariant' => $this->iriConverter->getItemIriFromResourceClass($variant::class, ['code' => $code]),
+            'productVariant' => sprintf('/api/v2/shop/product-variants/%s', $code),
```
Comment (TheMilek) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:1165]: same here

File: src/Sylius/Behat/Context/Setup/CartContext.php [hunk 22]
```diff
@@ -141,19 +143,21 @@ private function pickupCart(): string
...
 
             if ($user instanceof ShopUserInterface) {
                 /** @var CustomerInterface $customer */
-                $customer = $user->getCustomer();
-                $commandPickupCart->setEmail($customer->getEmail());
+                $email = $user->getCustomer()->getEmail();
             }
         }
 
+        $commandPickupCart = new PickupCart(
```
Comment (TheMilek) [inline src/Sylius/Behat/Context/Setup/CartContext.php]:
```suggestion
        $pickupCart = new PickupCart(
```
just wondering becouse above there is also `$updateCart` rather than `$commandUpdateCart`

File: src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChanger.php [hunk 23]
```diff
@@ -31,7 +31,7 @@ public function __construct(
 
     public function changePaymentMethod(
         string $paymentMethodCode,
-        string $paymentId,
+        int $paymentId,
```
Comment (TheMilek) [inline src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChanger.php:34]:
```suggestion
        int|string $paymentId,
```
I'm not really sure about that lets make it like that or even `mixed`

File: src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChangerInterface.php [hunk 24]
```diff
@@ -19,7 +19,7 @@ interface PaymentMethodChangerInterface
 {
     public function changePaymentMethod(
         string $paymentMethodCode,
-        string $paymentId,
+        int $paymentId,
```
Comment (TheMilek) [inline src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChangerInterface.php:22]:
```suggestion
        int|string $paymentId,
```
same here

File: src/Sylius/Bundle/ApiBundle/Resources/config/services.xml [hunk 25]
```diff
@@ -62,12 +62,12 @@
 <!--            <argument type="service" id="sylius.context.channel" />-->
 <!--        </service>-->
 
-<!--        <service id="Sylius\Bundle\ApiBundle\EventListener\ApiCartBlamerListener">-->
-<!--            <argument type="service" id="sylius.context.cart" />-->
-<!--            <argument type="service" id="sylius.section_resolver.uri_based_section_resolver" />-->
-<!--            <argument type="service" id="sylius.command_bus" />-->
-<!--            <tag name="kernel.event_listener" event="Symfony\Component\Security\Http\Event\LoginSuccessEvent" method="onLoginSuccess" />-->
-<!--        </service>-->
+        <service id="Sylius\Bundle\ApiBundle\EventListener\ApiCartBlamerListener">
```
Comment (TheMilek) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services.xml]: shouldn't it be dotted?

File: tests/Api/Responses/shop/checkout/completion/order_with_free_non_shippable_items.json [hunk 26]
```diff
@@ -18,7 +22,11 @@
         "@type": "Address",
         "firstName": "John",
         "lastName": "Doe",
+        "phoneNumber": null,
```
Comment (TheMilek) [inline tests/Api/Responses/shop/checkout/completion/order_with_free_non_shippable_items.json:25]: maybe we should use `skip_null_values` here and in other responses 🤔

File: src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Shop/Order/StateBasedExtension.php
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Shop/Order/StateBasedExtension.php:1]: The state is somewhat hardcoded. How about naming it something like `CartStateExtension` or something similar?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/OrderItem.xml [hunk 27]
```diff
@@ -29,17 +29,20 @@
...
             </operation>
+        </operations>
+    </resource>
 
-            <operation
-                class="ApiPlatform\Metadata\Get"
-                uriTemplate="/shop/order-items/{id}"
-                provider="sylius_api.state_provider.shop.order_item.item"
-            >
+    <resource class="Sylius\Component\Core\Model\OrderItem" uriTemplate="/shop/orders/{tokenValue}/items">
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/OrderItem.xml:35]: I have a question. In this case, the `<resource />` tag defines the same resource twice. Is there a difference between using a single resource tag and setting the `uriTemplate` in the operation tag?

File: tests/Api/Responses/shop/checkout/completion/order_with_free_non_shippable_items.json (hunk 26 above)
Comment (Rafikooo) [reply tests/Api/Responses/shop/checkout/completion/order_with_free_non_shippable_items.json:25]: IMHO, it's okay to return this nullable data. It's more concise and informative to know that the field is serialized but not set yet. Any other opinions?

File: tests/Api/Shop/OrdersTest.php [hunk 28]
```diff
@@ -25,11 +28,616 @@ final class OrdersTest extends JsonApiTestCase
...
     protected function setUp(): void
     {
+        $this->setUpDefaultGetHeaders();
         $this->setUpOrderPlacer();
 
         parent::setUp();
     }
 
+    /** @test */
+    public function it_gets_orders(): void
```
Comment (Rafikooo) [inline tests/Api/Shop/OrdersTest.php]:
```suggestion
    public function it_gets_only_logged_in_user_orders_excluding_guest_and_other_users_orders(): void
```
A bit long, but explicitly descriptive.

File: tests/Api/Shop/OrderItemsTest.php [hunk 29]
```diff
@@ -31,7 +31,142 @@ protected function setUp(): void
...
+        );
+
+        $this->assertResponse(
+            $this->client->getResponse(),
+            'shop/order_item/get_order_items_by_user',
+        );
+    }
+
+    /** @test */
+    public function it_gets_order_items_created_by_a_user_authenticated_as_another_user(): void
```
Comment (Rafikooo) [inline tests/Api/Shop/OrderItemsTest.php:61]: Why do we consider a case where another user's resource is accessed? If it's a valid scenario, it would be good to perform two requests: one as the user who placed the order, and another as a different user retrieving an empty array.

File: tests/Api/Responses/shop/order_item/get_order_item_by_guest.json [hunk 30]
```diff
@@ -1,6 +1,6 @@
 {
     "@context": "\/api\/v2\/contexts\/OrderItem",
-    "@id": "\/api\/v2\/shop\/order-items\/@integer@",
+    "@id": "\/api\/v2\/shop\/orders\/token\/items\/@integer@",
```
Comment (Rafikooo) [inline tests/Api/Responses/shop/order_item/get_order_item_by_guest.json:3]: Below, there is an `order` key, that could be changed from:
`"order": "\/api\/v2\/shop\/orders\/@string@"`
to
`"order": "\/api\/v2\/shop\/orders\/token"`.

File: tests/Api/Shop/Checkout/PaymentMethodTest.php [hunk 31]
```diff
@@ -0,0 +1,299 @@
...
+        $this->client->request(
+            method: 'PATCH',
+            uri: sprintf('/api/v2/shop/orders/%s/payments/%s', $tokenValue, $cart->getLastPayment()->getId()),
+            server: $this->headerBuilder()->withMergePatchJsonContentType()->withJsonLdAccept()->withShopUserAuthorization('dave@doe.com')->build(),
+            content: json_encode([
+                'paymentMethod' => $paymentMethod->getCode(),
+            ], \JSON_THROW_ON_ERROR),
+        );
+
+        $this->assertResponseCode($this->client->getResponse(), Response::HTTP_NOT_FOUND);
```
Comment (Rafikooo) [inline tests/Api/Shop/Checkout/PaymentMethodTest.php:101]:
```suggestion
        $this->assertResponseCode($this->client->getResponse(), Response::HTTP_FORBIDDEN);
```
Just opening a discussion. In another test, in a similar case where one user tries to retrieve another user's resource, we return a forbidden response, which makes sense. It would be good to take the same approach for all these similar cases.

File: tests/Api/Responses/shop/order_item/get_empty_order_items.json [hunk 32]
```diff
@@ -1,6 +1,6 @@
 {
   "@context": "\/api\/v2\/contexts\/OrderItem",
```
Comment (Rafikooo) [inline tests/Api/Responses/shop/order_item/get_empty_order_items.json]: Indentation issue.

File: tests/Api/Responses/shop/order_item/get_order_items_by_user.json [hunk 33]
```diff
@@ -0,0 +1,21 @@
+{
+  "@context": "\/api\/v2\/contexts\/OrderItem",
```
Comment (Rafikooo) [inline tests/Api/Responses/shop/order_item/get_order_items_by_user.json]: Indentation issue.

File: tests/Api/Responses/shop/checkout/completion/order_with_free_non_shippable_items.json (hunk 26 above)
Comment (Wojdylak) [reply tests/Api/Responses/shop/checkout/completion/order_with_free_non_shippable_items.json:25]: https://github.com/Sylius/Sylius/pull/16681#discussion_r1705716320

File: tests/Api/Shop/OrderItemsTest.php (hunk 29 above)
Comment (Wojdylak) [reply tests/Api/Shop/OrderItemsTest.php:61]: The above test checks whether a user who placed the order can retrieve an array of order items.

File: features/shop/checkout/ability_to_change_email_during_checkout.feature (hunk 17 above)
Comment (Wojdylak) [reply features/shop/checkout/ability_to_change_email_during_checkout.feature:14]: This suit will be enabled in the next PR

File: src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChanger.php (hunk 23 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChanger.php:34]: But I changed the variable in the command from string to int, so it will always be int.

File: src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChangerInterface.php (hunk 24 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChangerInterface.php:22]: https://github.com/Sylius/Sylius/pull/16692#discussion_r1721270285

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php (hunk 6 above)
Comment (Wojdylak) [reply src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: I'm not sure, maybe I did some tests and didn't revert it back to POST

File: src/Sylius/Bundle/ApiBundle/Command/Checkout/CompleteOrder.php (hunk 9 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/Command/Checkout/CompleteOrder.php:26]: I prefer to use a method rather than getting a value from a property, but we can see what other will say.

File: src/Sylius/Bundle/ApiBundle/Command/Checkout/ChoosePaymentMethod.php (hunk 10 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/Command/Checkout/ChoosePaymentMethod.php:27]: In your example, it's unclear which argument is from the body. I also think it would be better to have a default value for the argument which will be set by the context builder.

File: src/Sylius/Bundle/ApiBundle/Command/PaymentIdAwareInterface.php (hunk 11 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/Command/PaymentIdAwareInterface.php:16]: I think that we should consider removing this interface or implement this logic

File: src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Shop/Order/StateBasedExtension.php
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/Doctrine/ORM/QueryExtension/Shop/Order/StateBasedExtension.php:1]: I wanted to be consistent with the Extension for the Admin section.

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/OrderItem.xml (hunk 27 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_platform/resources/OrderItem.xml:35]: IMO this is more readable

File: tests/Api/Shop/Checkout/PaymentMethodTest.php (hunk 31 above)
Comment (Wojdylak) [reply tests/Api/Shop/Checkout/PaymentMethodTest.php:101]: In your case, as a guest, if you put in the wrong token, you get a NOT_FOUND error. However, if you enter the correct token, you get a FORBIDDEN error, which reveals that the token exists in the database.

File: src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChanger.php (hunk 23 above)
Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/Changer/PaymentMethodChanger.php:34]: The `id` should not be typehinted only to `int` also in command

File: src/Sylius/Bundle/ApiBundle/Command/Account/ChangePaymentMethod.php (hunk 15 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Command/Account/ChangePaymentMethod.php:25]:
```suggestion
        protected string $paymentMethodCode,
        protected mixed $paymentId,
        protected string $orderTokenValue,
```

File: src/Sylius/Bundle/ApiBundle/Command/Checkout/ChoosePaymentMethod.php [hunk 34]
```diff
@@ -15,64 +15,30 @@
...
-     */
-    public $paymentMethodCode;
-
-    public function __construct(string $paymentMethodCode)
-    {
-        $this->paymentMethodCode = $paymentMethodCode;
+    public function __construct(
+        public string $paymentMethodCode,
+        public ?int $paymentId = null,
+        public ?string $orderTokenValue = null,
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Command/Checkout/ChoosePaymentMethod.php:26]:
```suggestion
        protected string $paymentMethodCode,
        protected mixed $paymentId,
        protected string $orderTokenValue,
```
as there is a possibility to create a command with all unnecessary data using context builder, we should use this and not allow the creation of an object with incomplete data

File: src/Sylius/Bundle/ApiBundle/Command/Checkout/CompleteOrder.php [hunk 35]
```diff
@@ -17,24 +17,14 @@
...
-
-    /** @var string|null */
-    public $notes;
-
-    public function __construct(?string $notes = null)
-    {
-        $this->notes = $notes;
+    public function __construct(
+        public ?string $notes = null,
+        public ?string $orderTokenValue = null,
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Command/Checkout/CompleteOrder.php:22]:
```suggestion
        protected ?string $notes = null,
        protected string $orderTokenValue,
```
and the order of arguments could be changed 🤔

File: src/Sylius/Bundle/ApiBundle/Command/PaymentIdAwareInterface.php (hunk 11 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Command/PaymentIdAwareInterface.php:16]: This interface should not be needed anymore
```suggestion
interface PaymentIdAwareInterface
```

File: tests/Api/Shop/OrdersTest.php [hunk 36]
```diff
@@ -25,11 +28,616 @@ final class OrdersTest extends JsonApiTestCase
...
+            content: json_encode([
+                'paymentMethod' => $paymentMethod->getCode(),
+            ], \JSON_THROW_ON_ERROR),
+        );
+
+        $this->assertResponseCode($this->client->getResponse(), Response::HTTP_NOT_FOUND);
+    }
+
+    /** @test */
+    public function it_update_payment_method_of_order_created_by_a_user_authenticated_as_a_user(): void
```
Comment (GSadee) [inline tests/Api/Shop/OrdersTest.php:344]:
```suggestion
    public function it_updates_payment_method_of_order_created_by_a_user_authenticated_as_a_user(): void
```

**Review by GSadee — APPROVED** [review]
As the PR is enormous and there is no blocker, I am merging it as it stands and the added comments will be resolved in the next PR 👍🏻
